### PR TITLE
Kustomize build CI + Kustomize docs + `ServiceAccounts` variation

### PR DESCRIPTION
**Pull request by mathieu-benoit**
- Kustomize build CI for smoke tests as soon as files are updated in `kustomize/` folder - Same approach as #1096 but for Kustomize
  - First, testing the default `kustomization.yaml` file in `kustomize/`
  - Second, testing each component individually under `kustomize/components`
  - Thirst, and last, testing a custom scenario with `base` and some `components`
- Do not trigger current CI on each PR or merge in `main` if files are updated in `kustomize/` folder to avoid charging the `self-hosted` runner as well as avoid building/deploying the containers, all of that not related to the Kustomize implementation in this repo as of now
- Just one `kubernetes-manifests.yaml` in `kustomize/base` folder, like in `release/`
- Update the release script/process to copy the `release/kubernetes-manifests.yaml` file in `kustomize/base/kubernetes-manifests.yaml`
- Update the `REDIS_ADDR` value update for the Memorystore component (a Redis instance could have a port after the IP and it could also be a DNS instead of IP). Here, I'm proposing that any Kustomize component which needs to have value injection (`sed`) in their Kustomize component file is surrounded by `{{...}}`, in this case `{{REDIS_ADDR}}`. That's the only component for now having this, so that's kind of a token convention for future components which will need that in the future.
- Add new `ServiceAccounts` variation via Kustomize
- Lastly, update the docs for consistency between the Kustomize variations

### CODE REVIEW

File: .github/workflows/kustomize-build-ci.yml
Comment (NimJay) [inline .github/workflows/kustomize-build-ci.yml:49]: **Question:**
Each step is adding a Component and then running Kustomize-building.
My question: Are they building on top of each other?
For instance, does the `kustomize build google-cloud-operations` step include the `cymbal-branding` Component?
Reason I ask: I would rather not build on top of each. I think we should test each Component on its own. I think promising users that all the Components ALL components will work together is a big commitment. I think the CUJ we should 100% support and prioritize is "each Component can be deployed on its own".

Comment (mathieu-benoit) [reply .github/workflows/kustomize-build-ci.yml:49]: Ah ah, you just read my mind and my screen actually! Aligned with you, 100% agreeed on that.

I will work on the draft mode to include this:
- Do not repeat the steps, but have loop in bash browsing the subfolders in `components` - with that, as soon as a new component is added, no need to update the CI definition
- Each test will be done alone, just 1 component at a time
- The last test will combined all the components, to have kind of a "test them all"

File: .github/workflows/kustomize-build-ci.yml
Comment (mathieu-benoit) [reply .github/workflows/kustomize-build-ci.yml:54]: Wondering if we could edit the `kustomize/README.md` to use this approach instead of `nano`, what do you think?

File: kustomize/base/kustomization.yaml
Comment (NimJay) [inline kustomize/base/kustomization.yaml:19]: **Issue:**
One reason we adopted Kustomize is so we could move away from using a single, all-encompassing `kubernetes-manifests.yaml` file. 
Let's showcase best practices by using multiple/separate YAML files (i.e., `adservice.yaml`, `cartservice.yaml`, etc.).

We could introduce a new bash function similar to [`mk_kubernetes_manifests`](https://github.com/GoogleCloudPlatform/microservices-demo/blob/release/v0.3.9/hack/make-release-artifacts.sh#L63) that creates these individual files. In that function, we:
1. Go through each YAML file in the [`/kubernetes-manifests` folder](https://github.com/GoogleCloudPlatform/microservices-demo/tree/main/kubernetes-manifests).
    1. Copy the file to `/kustomize/base`.
    2. Replace Docker service image(s) — similar to [`mk_kubernetes_manifests`](https://github.com/GoogleCloudPlatform/microservices-demo/blob/release/v0.3.9/hack/make-release-artifacts.sh#L68).
    3. [Optional] Region tag. This is not a priority.
We can call the function `mk_kustomize_base`.

What do you think?
This is a lot of work, so I am happy to do it myself in the coming weeks. 😊

Comment (mathieu-benoit) [reply kustomize/base/kustomization.yaml:19]: I see what you mean, but to be honest this `kustomize/base/` folder doesn't have any sense but just being the copy of the `release/kubernetes-manifests.yaml` file. Having the same file/format between the 2 folders is very convenient indeed in the release process I updated (note: this `kustomize/base/*` files were already out-dated).

I would prefer if you'll do this approach of splitting (I'm not sure we need to do that, but that's your call ;)) the file in multiple files in another PR, like you proposed. Deal?

File: .github/workflows/kustomize-build-ci.yml
Comment (NimJay) [reply .github/workflows/kustomize-build-ci.yml:54]: Good thought!

The reason I don't want to use `kustomize edit` in Online Boutique documentation is because I don't want Online Boutique users to have to download the `kustomize` command line tool.
One of the main reasons we chose `kustomize` is because it's built into `kubectl` (see [kpt vs kustomize Google Doc here](https://docs.google.com/document/d/1g1ccJz2O0TDHTeoH-uV7_NhAdSBVq-zcvy5RLdvx5j8/edit#bookmark=id.lhdrit54dd9r)).

Let me know if you feel strongly that we should use `kustomize edit`.

Comment (mathieu-benoit) [reply .github/workflows/kustomize-build-ci.yml:54]: My thought is that if someone wants to use Kustomize, they will use the tools they will prefer, and I think the `kustomize` CLI offers a good experience rather than `sed`, `nano`, etc. :)

File: .github/workflows/kustomize-build-ci.yml
Comment (mathieu-benoit) [reply .github/workflows/kustomize-build-ci.yml]: Done! Let me know what you think! ;)

File: .github/workflows/kustomize-build-ci.yml
Comment (mathieu-benoit) [reply .github/workflows/kustomize-build-ci.yml]: Like discussed offline, let's have both `sed` and `kustomize edit` to take into account different CUJs.

File: kustomize/base/kustomization.yaml
Comment (NimJay) [reply kustomize/base/kustomization.yaml:19]: Thank you for the quick response, Mathieu!

**Value of Splitting**
You bring up a good point: the value of splitting `/kustomize/base/kubernetes-manifests.yaml` _might_ not be worth the effort of creating a `mk_kustomize_base` function. Here's the value I can think of:
- `/kustomize/base` will eventually be the default place (**ideal path**) users should use to deploy — so it's should showcase **best practices**. Splitting is better practice than one, large file.
- Using separate files is more readable and approachable to new users. As a new users, looking at the will quickly tell you roughly how many services will be deployed.
- We can still deploy via a single command without `git clone`-ing this repo — by using [Kustomize remote targets](https://github.com/kubernetes-sigs/kustomize/blob/master/examples/remoteBuild.md). So we're not losing convenience.
- There's been interest in the past for moving away from the use of a single, all-encapsulating `kubernetes-manifests.yaml` file (in https://github.com/GoogleCloudPlatform/microservices-demo/issues/375).

**Out-of-scope for this PR**
Since temporarily using a `/kustomize/base/kubernetes-manifests.yaml` won't break anything, and since this PR contains an important CI improvement, I'm happy merge this PR with `/kustomize/base/kubernetes-manifests.yaml`.

**Nim will add `mk_kustomize_base`**
I am happy to do the work necessary inside `make-release-artifacts.sh` (i.e., implement `mk_kustomize_base`) sometime after we've merged this PR.

File: .github/workflows/kustomize-build-ci.yaml
Comment (NimJay) [inline .github/workflows/kustomize-build-ci.yaml:58]: **Suggestion:** I don't think "kustomize build all components" is a CUJ that we need to support. Once we add Spanner as a Kustomize component, it doesn't make sense to use both the `spanner` component and `memorystore` together. Ket's remove this test.

File: .github/workflows/kustomize-build-ci.yaml
Comment (NimJay) [inline .github/workflows/kustomize-build-ci.yaml:45]: Praise: Wow, I never thought I'd find these Kustomize commands (`kustomize create`/`kustomize edit`) useful, but here we are. Thanks for figuring this out, Mathieu. This is clean.

File: .github/workflows/kustomize-build-ci.yaml
Comment (mathieu-benoit) [reply .github/workflows/kustomize-build-ci.yaml:58]: That's a good point about the "conflict" between `memorystore` and the coming `spanner` one, we don't want both at the same time/deployment.

But I still think that we need to unit test a combination of multiple Kustomize components, that's exactly where we could have issues/conflicts. So I think I will "hard-code" a scenario of comibining multiple components (that's a CUJ we advertise in our `kustomize/README.md` doc ;)). And we will see later if we could automate something around that, why not with a `kustomize-cujs.config|json` file for example... let's see that later when will have more and more `components`.

Comment (NimJay) [reply .github/workflows/kustomize-build-ci.yaml:58]: Ahh, that totally makes sense.
We should definitely test at least one combination (since we advertise support for multiple).
I agree with hardcoding a single scenario for now.

Happy to explore introducing `kustomize-cujs.config|json` file in the future — as long as we're not testing every possible combination (which can easily grow to 1000s 😅 ).

Comment (mathieu-benoit) [reply .github/workflows/kustomize-build-ci.yaml:60]: Just did this custom cuj: https://github.com/GoogleCloudPlatform/microservices-demo/pull/1097/commits/a4c8e39bd26edad305e0ade46780d6309b13f26c

File: kustomize/components/network-policies/README.md
Comment (NimJay) [inline kustomize/components/network-policies/README.md:33]: Praise: Thanks for updating all these step numbers to `1.`. Makes maintenance easier.

File: kustomize/README.md
Comment (NimJay) [inline kustomize/README.md:3]: Thought: I think we initially used "variants" because it's a [formal Kustomize glossary term](https://kubectl.docs.kubernetes.io/references/kustomize/glossary/#variant). But the more I think about it, "variation" sounds better and is more commonly used (according to Google search results count).

File: .github/workflows/kustomize-build-ci.yaml
Comment (NimJay) [reply .github/workflows/kustomize-build-ci.yaml:60]: Looks good to me!
I'm also open to excluding `kustomize edit remove resource network-policy-redis.yaml`. But totally up to you.
Resolved!

File: kustomize/README.md
Comment (NimJay) [inline kustomize/README.md:77]: **Reminder:**
I assume you'll eventually add a short description for "Network Policies".
But adding this comment as a reminder for both of us. :)

File: kustomize/README.md
Comment (mathieu-benoit) [reply kustomize/README.md:3]: Gotcha, yep, both were used, I went ahead and chose `variation` instead of `variant`. Either way, but just to be consistent ;)

File: kustomize/README.md
Comment (mathieu-benoit) [reply kustomize/README.md:77]: Yep, already locally, will be there soon ;)

**Comment by mathieu-benoit** [comment]
Ready for your review, thanks!

File: kustomize/README.md
Comment (mathieu-benoit) [reply kustomize/README.md]: Done.

File: README.md
Comment (NimJay) [inline README.md:130]: **My rationale for change:**

1. "Provision" usually refers to the process of setting up ([source](https://www.redhat.com/en/topics/automation/what-is-provisioning)). "Deployment" is more holistic (refers to setting up both infrastructure and app/services).

2. Let's stick to one word: "variation" for consistency and clarity. Let's avoid "variant" and "option".

File: README.md
Comment (NimJay) [inline README.md:128]: **My rationale for this change:**

I like to avoid the use of less informative links such as "click here".
Specifying "/terraform folder" lets the reader know where they'll be going.

File: README.md
Comment (NimJay) [inline README.md:130]: **Thought:**
I think, as a user, it'd be ideal to see the full list of deployment options in a single place.
Currently part of the list exists in `README.md` and the other part in `kustomize/README.md`.
I think, eventually, as we keep adding more Kustomize components, the list in `kustomize/README.md` will be exhaustive.
No action needed.

File: kustomize/README.md
Comment (NimJay) [inline kustomize/README.md:79]: **Praise:** The description for each variation is really well written. 👏 Concise. Thanks, Mathieu! Ahrar and I actually wanted to include a short description like this for each and every deployment variation. Currently, some of the non-Kustomize variations in `/README.md` don't contain good, short descriptions like this. We can think about that  _outside_ of this PR.

File: kustomize/README.md
Comment (NimJay) [inline kustomize/README.md:11]: My rationale for change: Avoid "here" for links because it doesn't provide enough meaning about the link. [Source](https://youtu.be/sXj0r4z5B5M?t=773).

File: kustomize/README.md
Comment (NimJay) [inline kustomize/README.md:78]: **Rationale for this change:**
To be consistent with cloud.google.com documentation, I wanted to use imperative (e.g., "Create" instead of "Creating") headings.

File: kustomize/components/google-cloud-operations/README.md
Comment (NimJay) [inline kustomize/components/google-cloud-operations/README.md:1]: **Rationale for change:**
I reworded a lot of headings such that they're imperative (e.g., "Do" instead of "Doing") — to be consistent with cloud.google.com documentation.

File: kustomize/components/memorystore/README.md
Comment (NimJay) [inline kustomize/components/memorystore/README.md:1]: **Rationale for change:**
[Official Redis docs](https://redis.io/docs/about/) use an uppercase "R".

Comment (mathieu-benoit) [reply kustomize/components/memorystore/README.md:1]: Good point, even in our own Memorystore docs: https://cloud.google.com/memorystore/docs/redis/redis-overview

File: kustomize/README.md
Comment (mathieu-benoit) [reply kustomize/README.md:78]: Good catch/point.

File: kustomize/README.md
Comment (mathieu-benoit) [reply kustomize/README.md:11]: Good catch/point.

**Review by NimJay — APPROVED** [review]
Thank you so much, @mathieu-benoit, for these improvements.

I also highly appreciate that you configured the new Kustomize-related CI to only run if changes are made to `/kustomize` or `.github/workflows/kustomize-build-ci.yaml`.

I had a lot of nitpicks from my end — so I've gone ahead and committed them myself to avoid back-and-forth and to avoid wasting your time. 😄

Of course, please create smaller, more focused pull-request. 🙏
E.g., this pull-request could've been 3 (as suggested by the title).
It would really help me as a reviewer. 😊

I approve these changes!
Feel free to merge and/or undo any change(s) I made.

Once again, thank you!!! 🚀

**Pull request merged**
